doc/2019-08-04-write-Amplification.md

```diff
@@ -0,0 +1,208 @@
+4. Incur unnecessary WA and affect performance during GC.
+
+## Compatibility
+
+It does not affect the compatibility with Linux. Unfortunately, the method, fallocate,  used in this proposal is limited to Linux 3.1, not to the broader system interface, such as POSIX.
+
+## Implementation
+
+### Dig Hole
+
+- Who: Hanyang Liu [lhy1024](https://github.com/lhy1024)
+- When: 7.23-8.12
+- Step 1: Get input files after sample and parallel dig holes in files.
+- Step 2: 
+  We will foreach through all the records in a file and save the ending offset of the last valid record. If a new valid record is found, we will dig holes between the end offset of the last valid record and the start offset of the record.
+  In addition to buffering the end offset of a valid record, we will also save the key of the previous record and whether it is valid, which will reduce access to the LSM-tree and reduce decoding with the record.
+
+### File operation support
+
+- Who: Wenjun Huang [WjHuang](https://github.com/wjhuang2016)
+- When: 7.23-8.12
+- Step 1: Implement a file class bases on PosixRandomRWFile provide by rocksdb and provides some functions, such as seek, punch_hole. 
+- Step 2: Rewriter the blob file iterator to support reading from a blob file with holes. 
+
+### Blob GC picker
+
+- Who: Ning Lin [NingLin-P](https://github.com/NingLin-P)
+- When: 7.23-8.12
+- Step 1: according to the discardable size of each blob files, choose some files to perform hole-punching, and also choose blob files with small size to perform compaction like Titan
+- Step 2: While in Titan blob files are immutable after writing finish and being selected to GC mean this blob file will be deleted soon, but in our approach this property no longer hold and hence we need to carefully rethink about the management of blob files, especially in case of concurrent operation.
+
+## Evaluation
+
+### Test Environment
+
+- CPU: I5-8350u
+- RAM: 24GB
+- SSD: SAMSUNG 970Pro 512GB 
+
+
+### Case 1: Different Record Size
+#### Workload
+
+- Operation: 100% update
+- Value size: from 1K to 512K
+- Record count: 4G/value size
+- Operation count: 16G/value size
+- Request distribution: uniform
+
+#### Result
+
+<p align="center">
+  <img src="./imgs/1565839903157.png"><br> 
+  Figure 2: Write Amplification
+</p>
+
+<p align="center">
+  <img src="./imgs/1565840445540.png"><br> 
+  Figure 3: Space Amplification
+</p>
+
+<p align="center">
+  <img src="./imgs/1565840407310.png"><br> 
+  Figure 4: OPS
+</p>
+
+#### Analysis
+
+We can see that in both cases of 2k and 4k, the use of hole punching instead introduces a larger write amplification. Because we use 4k alignment to process data, which leads to the storage of one such record need more blocks to be filled zero. And the data we write is not exactly 2k or 4k, considering the head and key. In fact, it is just a little bigger than 2k or 4k. 
```

Comment:
Just to put a note here. To avoid the need to align records, we can have a key->offset map at the end of each blob file, and have GC job iterate through this map instead. The map can probably also reduce GC read amp.

Reply:
Yes, we also had have such a plan, but at the time we only had a few days left for talent plan, so we chose a method that was easier to implement.
However, misalignment may result in a lot less space for punching holes, because cross-block records will be more.